llvm-general/src/LLVM/General/Internal/FFI/OrcJIT.hpp

```diff
@@ -0,0 +1,59 @@
+#ifndef __LLVM_GENERAL_INTERNAL_FFI__ORC_JIT__HPP__
+#define __LLVM_GENERAL_INTERNAL_FFI__ORC_JIT__HPP__
+
+#include "LLVM/General/Internal/FFI/OrcJIT.h"
+#include "LLVM/General/Internal/FFI/Target.hpp"
+
+#include "llvm/ExecutionEngine/Orc/IRCompileLayer.h"
+#include "llvm/ExecutionEngine/Orc/JITSymbol.h"
+#include "llvm/ExecutionEngine/Orc/ObjectLinkingLayer.h"
+
+typedef llvm::orc::ObjectLinkingLayer<> *LLVMObjectLinkingLayerRef;
+typedef llvm::orc::IRCompileLayer<llvm::orc::ObjectLinkingLayer<>>
+    *LLVMIRCompileLayerRef;
+typedef llvm::orc::JITSymbol *LLVMJITSymbolRef;
+typedef llvm::orc::ObjectLinkingLayer<>::ObjSetHandleT *LLVMModuleSetHandleRef;
+
+extern "C" {
+LLVMIRCompileLayerRef
+LLVM_General_createIRCompileLayer(LLVMObjectLinkingLayerRef objectLayer,
+                                  LLVMTargetMachineRef tm);
```

Comment:
I think it's unlikely that these functions will ever be called by anything other than Haskell, including each other, I think both of the usual reasons for declaring them in headers don't apply. Please do take out what declarations you can and of those which remain (including the type declarations) move those which you can into the cpp file.